src/sentry_core.c

```diff
@@ -934,6 +934,9 @@ sentry_transaction_finish(sentry_transaction_t *opaque_tx)
     sentry_value_t trace_context
         = sentry__value_get_trace_context(opaque_tx->inner);
     sentry_value_t contexts = sentry_value_new_object();
+    sentry_value_set_by_key(
+        trace_context, "data", sentry_value_get_by_key(tx, "data"));
+    sentry_value_incref(sentry_value_get_by_key(tx, "data"));
```

Comment:
There is one missing piece left (added a todo to the PR description): we should apply the currently scoped transaction/span "data" to every event sent inside that transaction/span here:

https://github.com/getsentry/sentry-native/blob/ffdf6ed024ff1f2fe2e55fbe7d86940ca0228207/src/sentry_scope.c#L321-L328

If the `scope_trace` isn't a null object, we can also read the "data" item of that scoped transaction/span and place it in the events trace context. Again, you must use `incref` here because it will be `decref`ed when sending the envelope, meaning it would be freed and no longer accessible to the transaction/span.

Removal is unnecessary since the transaction/span will continue to exist beyond the lifetime of the event.

Reply:
The data item should only be added to the event if it isn't a null object in the scoped transaction/span. It doesn't make a difference in the backend (null object == non-existent key, most of the time), but there is no need to send data over the wire that signifies non-existent.

Reply:
Is this a correct implementation of what you asked for? (i.e. when a transaction object exists inside the scope, we get its "data" key item and put it at the scope_trace "data" key)
```c
sentry_value_t scope_trace = sentry__value_get_trace_context(
    sentry__get_span_or_transaction(scope));
if (!sentry_value_is_null(scope_trace)) {
    if (sentry_value_is_null(contexts)) {
        contexts = sentry_value_new_object();
    }
    if (scope->transaction_object) {
        sentry_value_set_by_key(scope_trace, "data",
            sentry_value_get_by_key(
                scope->transaction_object->inner, "data"));
        sentry_value_incref(sentry_value_get_by_key(
            scope->transaction_object->inner, "data"));
    }
    sentry_value_set_by_key(contexts, "trace", scope_trace);
}
```

Reply:
You can use either `span` or `transaction` as returned from `sentry__get_span_or_transaction(scope)` and don't have to retrieve the transaction specifically. Both can have `data` items, and they should apply theirs to the event. If, for instance, a span should inherit its `data` from the transaction, then that should happen at span creation.

@markushi, do you also apply `data` when you apply the trace context to events from your scope transaction/spans? If so, do you take it solely from the transaction or also from spans?

Reply:
To clarify, I would probably do something like this:

```C
sentry_value_t scoped_txn_or_span = sentry__get_span_or_transaction(scope)
sentry_value_t scope_trace = sentry__value_get_trace_context(scoped_txn_or_span);
if (!sentry_value_is_null(scope_trace)) {
    if (sentry_value_is_null(contexts)) {
        contexts = sentry_value_new_object();
    }
    sentry_value_t scoped_txn_or_span_data = sentry_value_get_by_key(scoped_txn_or_span, "data");
    if (!sentry_value_is_null(scoped_txn_or_span_data)) {
        sentry_value_incref(scoped_txn_or_span_data);
        sentry_value_set_by_key(scope_trace, "data", scoped_txn_or_span_data);
    }
    sentry_value_set_by_key(contexts, "trace", scope_trace);
}
```

Reply:
@supervacuus @JoshuaMoelans Yes, we do the same but we solely take the context from the transaction:
https://github.com/getsentry/sentry-java/blob/28a11a7925b8e45bec38684f2002192787f13477/sentry/src/main/java/io/sentry/SentryClient.java#L216-L229

It's worth mentioning that at least in `sentry-java` a span doesn't really have it's own trace context, it solely returns the top level txn context (see https://github.com/getsentry/sentry-java/blob/2e90ac7733c6c9b941e7c64175f3d097d1d15a37/sentry/src/main/java/io/sentry/Span.java#L170)